feat: add spfx utils for disambiguation (#1914)
microsoftgraph/microsoft-graph-toolkit: packages/mgt-components/README.md

    <mgt-agenda group-by-day></mgt-agenda>
    ```

## <a id="disambiguation">Disambiguation</a>

MGT is built using [web components](https://developer.mozilla.org/en-US/docs/Web/Web_Components). Web components use their tag name as a unique key when registering within a browser. Any attempt to register a component using a previously registered tag name results in an error being thrown when calling `CustomElementRegistry.define()`. In scenarios where multiple custom applications can be loaded into a single page this created issues for MGT, most notably in developing solutions using SharePoint Framework.

> To mitigate this challenge we built the [`mgt-spfx`](https://github.com/microsoftgraph/microsoft-graph-toolkit/tree/main/packages/mgt-spfx) package. Using `mgt-spfx` developers could centralize the registration of MGT web components across all SPFx solutions deployed on the tenant. By reusing MGT components from a central location web parts from different solutions could be loaded into a single page without throwing errors. When using `mgt-spfx` all MGT based web parts in a SharePoint tenant used the same version of MGT.

Reviewer: If we "deprecate" the `mgt-spfx` package, will it stay to that location?

Author: Good question. I've been thinking that we might not need to deprecate `mgt-spfx` as it still provides value even with disambiguation. And hey one day we might be able to deliver mgt-spfx@4 with a disambiguation baked in to run alongside v2 or v3

Author: @sebastienlevert I changed the tense used here to indicate that `mgt-spfx` is still an option, marking as resolved

To allow developers to build web parts using the latest version of MGT and load them on pages along web parts that use v2.x of MGT, we've added a new disambiguation feature to MGT. Using this feature developers can specify a unique string to add to the tag name of all MGT web components in their application.

### Usage in standard HTML and JavaScript

The earlier example can be updated to use the disambiguation feature as follows:

```html
<script type="module">
  import { Providers, customElementHelper } from '@microsoft/mgt-element';
  customElementHelper.withDisambiguation('contoso');
  import { Msal2Provider } from '@microsoft/mgt-msal2-provider';
  Providers.globalProvider = new Msal2Provider({clientId: 'clientId'});

</script>

<mgt-contoso-login></mgt-contoso-login>
<mgt-contoso-person person-query="Bill Gates" person-card="hover"></mgt-contoso-person>
<mgt-contoso-agenda group-by-day></mgt-contoso-agenda>
```

> > Note: the `import` of `mgt-components` must use a dynamic import to ensure that the disambiguation is applied before the components are imported.

Reviewer: In this sample, I don't see the `mgt-components` being loaded dynamically. Should we have that in there also?

Author: lol, good catch!

A complete example is available in the [React SharePoint Web Part Sample](https://github.com/microsoftgraph/microsoft-graph-toolkit/blob/main/samples/sp-webpart/src/webparts/mgtDemo/MgtDemoWebPart.ts).

```ts
// [...] trimmed for brevity
import { Providers } from '@microsoft/mgt-element/dist/es6/providers/Providers';
import { customElementHelper } from '@microsoft/mgt-element/dist/es6/components/customElementHelper';
import { SharePointProvider } from '@microsoft/mgt-sharepoint-provider/dist/es6/SharePointProvider';
import { lazyLoadComponent } from '@microsoft/mgt-spfx-utils';

// Async import of component that imports the React Components
const MgtDemo = React.lazy(() => import('./components/MgtDemo'));

export interface IMgtDemoWebPartProps {
  description: string;
}
// set the disambiguation before initializing any webpart
customElementHelper.withDisambiguation('bar');

export default class MgtDemoWebPart extends BaseClientSideWebPart<IMgtDemoWebPartProps> {
  // set the global provider
  protected async onInit() {
    if (!Providers.globalProvider) {
      Providers.globalProvider = new SharePointProvider(this.context);
    }
  }

  public render(): void {

>     const element = lazyLoadComponent(MgtDemo, { description: this.properties.description });

Reviewer: This is absolutely brilliant! 🎉

    ReactDom.render(element, this.domElement);
  }

  // [...] trimmed for brevity
}
```

### Dynamic imports aka Lazy Loading

Using dynamic imports you can load dependencies asynchronously. This pattern allows you to load dependencies only when needed. For example, you may want to load a component only when a user clicks a button. This is a great way to reduce the initial load time of your application. In the context of disambiguation, you need to use this technique because components register themselves in the browser when they are imported.

**Important:** If you import the components before you have applied the disambiguation, the disambiguation will not be applied and using the disambiguated tag name will not work.

When using an `import` statement the import statement is hoisted and executed before any other code in the code block. To use dynamic imports you must use the `import()` function. The `import()` function returns a promise that resolves to the module. You can also use the `then` method to execute code after the module is loaded and the `catch` method to handle any errors if necessary.

```typescript
// static import via a statement

> import '@microsoft/mgt-components';
> 
> // dynamic import via a function
> import('@microsoft/mgt-components').then(() => {
>   // code to execute after the module is loaded
> }).catch((e) => {
>   // handle any errors
> });
> ```

Author: ```suggestion
import { Providers, customElementHelper } from '@microsoft/mgt-element';
import { Msal2Provider } from '@microsoft/mgt-msal2-provider';

customElementHelper.withDisambiguation('contoso');
Providers.globalProvider = new Msal2Provider({clientId: 'clientId'});

// dynamic import via a function
import('@microsoft/mgt-components').then(() => {
  // code to execute after the module is loaded
}).catch((e) => {
  // handle any errors
});
```

To mitigate this challenge we built the [`mgt-spfx`](https://github.com/microsoftgraph/microsoft-graph-toolkit/tree/main/packages/mgt-spfx) package. Using `mgt-spfx` developers could centralize the registration of MGT web components across all SPFx solutions deployed on the tenant. By reusing MGT components from a central location web parts from different solutions could be loaded into a single page without throwing errors. When using `mgt-spfx` all MGT based web parts in a SharePoint tenant used the same version of MGT.

To allow developers to build web parts using the latest version of MGT and load them on pages along web parts that use v2.x of MGT, we've added a new disambiguation feature to MGT. Using this feature developers can specify a unique string to add to the tag name of all MGT web components in their application.

### Usage in standard HTML and JavaScript

The earlier example can be updated to use the disambiguation feature as follows:

```html
<script type="module">

>   import { Providers, customElementHelper } from '@microsoft/mgt-element';
>   customElementHelper.withDisambiguation('contoso');
>   import { Msal2Provider } from '@microsoft/mgt-msal2-provider';
>   Providers.globalProvider = new Msal2Provider({clientId: 'clientId'});

Author: ```suggestion
  import { Providers, customElementHelper } from '@microsoft/mgt-element';
  import { Msal2Provider } from '@microsoft/mgt-msal2-provider';
  customElementHelper.withDisambiguation('contoso');
  Providers.globalProvider = new Msal2Provider({clientId: 'clientId'});
  
  // import the components using dynamic import to avoid hoisting
  import('@microsoft/mgt-components');
```

**Important:** If you import the components before you have applied the disambiguation, the disambiguation will not be applied and using the disambiguated tag name will not work.

When using an `import` statement the import statement is hoisted and executed before any other code in the code block. To use dynamic imports you must use the `import()` function. The `import()` function returns a promise that resolves to the module. You can also use the `then` method to execute code after the module is loaded and the `catch` method to handle any errors if necessary.

```typescript
// static import via a statement
import { Providers, customElementHelper } from '@microsoft/mgt-element';
import { Msal2Provider } from '@microsoft/mgt-msal2-provider';

customElementHelper.withDisambiguation('contoso');
Providers.globalProvider = new Msal2Provider({clientId: 'clientId'});

// dynamic import via a function
import('@microsoft/mgt-components').then(() => {

>   // code to execute after the module is loaded

Author: I've assumed that you were meaning to add a separate static import example as we can't mix static and dynamic imports of `mgt-components` as it will break the application.

>   // code to execute after the module is loaded

Author: @waldekmastykarz if you could let me know if this hits the mark or not that would be fantastic.

>   // code to execute after the module is loaded

Reviewer 2: Right, sorry for not being clearer. What I meant was to extend the sample with an example showing for example `<mgt-person/>` so that readers know where the contents of their app should go. You bring up a good point that it's not possible to mix both approach in a single code snippet.

>   // code to execute after the module is loaded

Author: Thank you for the clarification, I think we're good on this one now

    <mgt-agenda group-by-day></mgt-agenda>
    ```

## <a id="disambiguation">Disambiguation</a>

> MGT is built using [web components](https://developer.mozilla.org/en-US/docs/Web/Web_Components). Web components use their tag name as a unique key when registering within a browser. Any attempt to register a component using a previously registered tag name results in an error being thrown when calling `CustomElementRegistry.define()`. In scenarios where multiple custom applications can be loaded into a single page this created issues for MGT, most notably in developing custom SharePoint web parts.

Reviewer 2: > custom SharePoint web parts

Let's generalize it to "developing solutions using SharePoint Framework." This issue applies to all UI elements that you can build with SPFx, such as web parts, extensions, field customizers, etc.

> To mitigate this challenge the [`mgt-spfx`](https://github.com/microsoftgraph/microsoft-graph-toolkit/tree/main/packages/mgt-spfx) package was built. This provided a library extension to SharePoint which centralized the registration of the MGT web components. This allowed for multiple MGT web parts to be loaded into a single page without throwing errors. When using `mgt-spfx` all MGT based web parts in a SharePoint tenant used the same version of MGT.

Reviewer 2: Let's switch to active voice to make it easier to read:

```suggestion
To mitigate this challenge we built the [`mgt-spfx`](https://github.com/microsoftgraph/microsoft-graph-toolkit/tree/main/packages/mgt-spfx) package. Using `mgt-spfx` developers could centralize the registration of MGT web components across all SPFx solutions deployed on the tenant. By reusing MGT components from a central location web parts from different solutions could be loaded into a single page without throwing errors. When using `mgt-spfx` all MGT based web parts in a SharePoint tenant used the same version of MGT.
```

> In order to allow developers to build custom MGT web parts using the latest version of MGT and retain the ability to have web parts built using v2.x of MGT on the same page a new disambiguation feature has been added to MGT. This allows developers to specify a unique string to add to the tag name of all MGT web components in their application.

Reviewer 2: ```suggestion
To allow developers to build web parts using the latest version of MGT and load them on pages along web parts that use v2.x of MGT, we've added a new disambiguation feature to MGT. Using this feature developers can specify a unique string to add to the tag name of all MGT web components in their application.
```

The earlier example can be updated to use the disambiguation feature as follows:

```html
<script type="module">
  import { Providers, customElementHelper } from '@microsoft/mgt-element';
  customElementHelper.withDisambiguation('contoso');
  import { Msal2Provider } from '@microsoft/mgt-msal2-provider';
  Providers.globalProvider = new Msal2Provider({clientId: 'clientId'});

</script>

<mgt-contoso-login></mgt-contoso-login>
<mgt-contoso-person person-query="Bill Gates" person-card="hover"></mgt-contoso-person>
<mgt-contoso-agenda group-by-day></mgt-contoso-agenda>
```

> > Note: the `import` of `mgt-components` must use a dynamic import to ensure that the disambiguation is applied before the components are imported. Helper utilities have been provided in the  [`mgt-spfx-utils`](https://github.com/microsoftgraph/microsoft-graph-toolkit/tree/main/packages/mgt-spfx-utils) package to make this easier in the context of SharePoint web parts.

Reviewer 2: Let's add an empty line between the code snippet and the callout

> > Note: the `import` of `mgt-components` must use a dynamic import to ensure that the disambiguation is applied before the components are imported. Helper utilities have been provided in the  [`mgt-spfx-utils`](https://github.com/microsoftgraph/microsoft-graph-toolkit/tree/main/packages/mgt-spfx-utils) package to make this easier in the context of SharePoint web parts.

Reviewer 2: > the `import` of `mgt-components` must use a dynamic import to ensure that the disambiguation is applied before the components are imported

This makes it sound like we omitted a part of code sample. Let's expand the existing sample or add a new one, so that all information that developers need is in one place.

> > Note: the `import` of `mgt-components` must use a dynamic import to ensure that the disambiguation is applied before the components are imported. Helper utilities have been provided in the  [`mgt-spfx-utils`](https://github.com/microsoftgraph/microsoft-graph-toolkit/tree/main/packages/mgt-spfx-utils) package to make this easier in the context of SharePoint web parts.

Reviewer 2: ```suggestion
> Note: to work properly, disambiguation must be applied before components are imported. This is why you need to use a dynamic import when referencing MGT components, when you use disambiguation. To make it easier in the context of SharePoint Framework solutions, we provide helper utilities in the  [`mgt-spfx-utils`](https://github.com/microsoftgraph/microsoft-graph-toolkit/tree/main/packages/mgt-spfx-utils) package.
```

### Dynamic imports aka Lazy Loading

> Dynamic imports are a means to asynchronously load code. This is useful when you want to load code only when it is needed. For example, you may want to load a component only when a user clicks a button. This is a great way to reduce the initial load time of your application. In the context of disambiguation it is necessary to use this technique as the components register themselves in the browser when they are imported.

Reviewer 2: ```suggestion
Using dynamic imports you can load dependencies asynchronously. This pattern allows you to load dependencies only when needed. For example, you may want to load a component only when a user clicks a button. This is a great way to reduce the initial load time of your application. In the context of disambiguation, you need to use this technique becasuse components register themselves in the browser when they are imported.
```

When using an `import` statement the import statement is hoisted and executed before any other code in the code block. To use dynamic imports you must use the `import()` function.

> 

Reviewer 2: Let's add a code sample where we show static import and its equivalent dynamic import, to help developers understand what we're talking about. Perhaps they already know the concept, but don't know that it's called _dynamic imports_ so a code sample could help them put the two together.

Author: Sample added and re-ordered the section slightly

Reviewer 2: Just to make it even clearer, shall we add some surrounding code to make it explicit that a static import goes at the top of the file, while a dynamic import is located somewhere in the code, where you want to load it explicitly? I think that would help folks who've never seen the pattern understand it better.

Author: Yeah that's a good idea. Although the import statement at the top of the file is a good practice due to the fact it gets hoisted, not a language requirement.

Author: @waldekmastykarz I've expanded this using our continued example

> If you import the components before you have applied the disambiguation the disambiguation will not have been applied and using the disambiguated tag name will not work.

Reviewer 2: Let's call this out:

```suggestion
**Important:** If you import the components before you have applied the disambiguation, the disambiguation will not be applied and using the disambiguated tag name will not work.
```
